Enable MSVC GPU jobs for CCCL projects

alliepiper opened the pull request:
## Summary

Enables Windows (MSVC) GPU CI jobs across CCCL projects: workflow action, PowerShell build/test scripts, and matrix compiler/GPU entries. Prerequisite fixes land first so each is independently bisectable.

**Prerequisite fixes:**
- Close temp file handle before removing it (`c/parallel`)
- Guard `__device__` lambdas with `EXTENDED_LAMBDA` in `for_each_in_extents` and `for_each_in_layout` tests (`cub`)
- Shorten long pstl scan test filenames (`libcudacxx`)
- Skip two failing tests on MSVC with issue refs — `no_unique_address` (#8953), `tuple_size_structured_bindings` (#8952) (`libcudacxx`)

**Windows CI infrastructure:**
- `.github/actions/workflow-run-job-windows/action.yml` — PowerShell runner action with colored output
- `ci/windows/build_common.psm1`, `test_libcudacxx.ps1`, `test_packaging.ps1` — build/test scripts
- `ci/matrix.yaml` — MSVC compiler entries (`msvc2019`, `msvc14.39`, `msvc2022`) across CTK versions; A100 GPU spec; cudax excluded until #8903 is resolved
- `cmake/CCCLUtilities.cmake` — `_CCCL_WSL` / `_CCCL_WINDOWS` detection helpers
- `cudax/test/CMakeLists.txt` — unified-memory tests guarded against Windows/WSL

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented on ci/windows/build_common.psm1:
```diff
@@ -1,4 +1,4 @@
-Param(
```
Diffs on these lines are due to BOM updates explicitly telling powershell that these are UTF-8 encoded.

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
Just switch to zstd compression for artifacts. Recording old measurements on CUBs lid1 artifact:

bzip2:
```
  Artifact 'z_cub-test-artifacts-cuda13.0-cl14.44-M-lid_1' packed in 581 seconds.
  Artifact 'z_cub-test-artifacts-cuda13.0-cl14.44-M-lid_1' fetched in 74 seconds.
  Artifact 'z_cub-test-artifacts-cuda13.0-cl14.44-M-lid_1' unpacked in 337 seconds.
  z_cub-test-artifacts-cuda13.0-cl14.44-M-lid_1 | 552 MB
```

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

miscco commented:
@pciolkosz I believe we need to disable some of the memory resource tests for MSVC https://github.com/NVIDIA/cccl/actions/runs/18978616950/job/54206675992?pr=5725#step:4:14641

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test f4ad6aa

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test

alliepiper commented:
/ok to test 76e6df1

alliepiper commented:
/ok to test 02e4eba

alliepiper commented:
/ok to test 568f5fd2b9

alliepiper commented:
/ok to test 150e711b97be9768bb562adf55354ce4736ceda3

alliepiper commented:
/ok to test 0082115c3989a6be63f774dbeeba2b510fc82e83

alliepiper committed:
Close temp file handle before removing it.

alliepiper committed:
Guard __device__ lambda with EXTENDED_LAMBDA in for_each_in_extents test.

alliepiper committed:
Guard __device__ lambda with EXTENDED_LAMBDA in for_each_in_layout test.

alliepiper committed:
Shorten long pstl scan test filenames.

alliepiper committed:
[libcu++] Skip two failing tests on MSVC (refs #8952, #8953)

alliepiper committed:
Enable Windows CI: workflow action, build/packaging scripts, matrix coverage.

alliepiper committed:
Exclude cudax tests on MSVC and WSL.

cudax tests fail intermittently under MSVC on Windows (tracked in #8903),
and the cudax unified-memory tests don't run reliably under WSL. Add a
matrix exclude entry for the former and a CMake-side guard for the latter.

alliepiper commented on ci/windows/test_packaging.ps1:
```diff
@@ -20,6 +23,20 @@ If($CURRENT_PATH -ne "ci") {
     pushd "$PSScriptRoot/.."
 }
 
+if ($MIN_CMAKE) {
+    $minVersion = "3.18.0"
+    Write-Host "Installing minimum CMake version v$minVersion..."
+    $url     = "https://github.com/Kitware/CMake/releases/download/v$minVersion/cmake-$minVersion-win64-x64.zip"
+    $zipPath = Join-Path $env:TEMP "cmake-min.zip"
+    $extract = Join-Path $env:TEMP "cmake-$minVersion"
+    Invoke-WebRequest -Uri $url -OutFile $zipPath -UseBasicParsing
+    if (Test-Path $extract) { Remove-Item -Recurse -Force $extract }
+    Expand-Archive -Path $zipPath -DestinationPath $extract -Force
+    $ctestExe = (Get-ChildItem -Path $extract -Recurse -Filter "ctest.exe" |
+                 Select-Object -First 1).FullName
+    $env:MIN_CTEST_COMMAND = $ctestExe
+}
```
not gonna happen irl

alliepiper commented on cub/test/catch2_test_device_for_each_in_layout_api.cu:
```diff
@@ -55,6 +52,7 @@ C2H_TEST("Device ForEachInLayout", "[ForEachInLayout][device]")
                                             {0, 1, 1}, {1, 1, 1}, {2, 1, 1}};
 
+#if __CUDACC_EXTENDED_LAMBDA__
   auto status = cub::DeviceFor::ForEachInLayout(mapping_left_type{extents},
```
Fixing.

alliepiper committed:
Restore missing test.

alliepiper committed:
Merge branch 'main' into windows_ci

alliepiper commented:
/ok to test 5881d28

alliepiper committed:
Merge remote-tracking branch 'origin/main' into windows_ci

alliepiper committed:
[libcu++] Re-enable layout_stride no_unique_address test on MSVC (#8953) [skip-vdc][skip-docs][skip-tpt]

alliepiper committed:
[libcu++] Skip complex tanh/tan tests on MSVC (#9166) [skip-vdc][skip-docs][skip-tpt]

alliepiper committed:
Merge remote-tracking branch 'origin/main' into windows_ci

alliepiper committed:
Apply suggestions from code review

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

alliepiper commented:
/ok to test 66c092a

alliepiper committed:
[libcudacxx] Suppress -Wno-attributes in nvcc lit test host compiler flags

alliepiper committed:
[libcudacxx] Fix __constant_wrapper forward decl for MSVC+NVCC

alliepiper committed:
Merge branch 'main' into windows_ci

alliepiper commented:
/ok to test e1378fd

alliepiper committed:
Merge branch 'main' into windows_ci

alliepiper commented:
/ok to test 99b61d4

miscco committed:
Merge branch 'main' into windows_ci

miscco commented on cmake/CCCLUtilities.cmake:
```diff
@@ -289,3 +289,17 @@ function(cccl_add_xfail_compile_target_test target_name)
+
+# Check for WSL. CMAKE_SYSTEM is "Linux-<kernelver>-microsoft-standard-WSL2" on WSLv2.
+if (CMAKE_SYSTEM MATCHES WSL)
+  set(_CCCL_WSL TRUE)
```
I would prefer if we prefixed with `_CCCL_CMAKE_WSL`

alliepiper replied:
I'm not sure what that adds -- there are CMake-specific implementation details, meant to be analogous to the bare global `WIN32` etc vars that CMake defines for platform detection. I'm not sure what benefit changing this would add -- all of these variables are cmake variables.

miscco commented on libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/tan.pass.cpp:
```diff
@@ -10,6 +10,9 @@
 
+// UNSUPPORTED: msvc
+// See https://github.com/NVIDIA/cccl/issues/9166
+
```
```suggestion
```

miscco commented on libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/tanh.pass.cpp:
```diff
@@ -10,6 +10,9 @@
 
+// UNSUPPORTED: msvc
+// See https://github.com/NVIDIA/cccl/issues/9166
+
```
```suggestion
```

miscco commented on libcudacxx/test/libcudacxx/std/utilities/tuple/tuple.tuple/tuple.helper/tuple_size_structured_bindings.pass.cpp:
```diff
@@ -6,6 +6,9 @@
 
+// UNSUPPORTED: msvc
+// See https://github.com/NVIDIA/cccl/issues/8952
+
```
```suggestion
```

miscco commented on libcudacxx/test/utils/libcudacxx/test/config.py:
```diff
@@ -1256,6 +1256,7 @@ def addIfHostSupports(flag):
                 addIfHostSupports("-Wno-deprecated-declarations")
                 addIfHostSupports("-Wno-noexcept-type")
                 addIfHostSupports("-Wno-unused-function")
+                addIfHostSupports("-Wno-attributes")
```
I believe this should stay, we should try and suppress this locally
```suggestion
```

miscco committed:
Apply suggestions from code review

Co-authored-by: Michael Schellenberger Costa <miscco@nvidia.com>

miscco commented on libcudacxx/include/cuda/std/__utility/constant_wrapper.h:
```diff
@@ -74,7 +74,7 @@ struct __cw_fixed_value<_Tp[_Extent]>
 _CCCL_DEDUCTION_GUIDE_ATTRIBUTES __cw_fixed_value(_Tp (&)[_Extent]) -> __cw_fixed_value<_Tp[_Extent]>;
 
-template <__cw_fixed_value _Xp, class = typename decltype(__cw_fixed_value(_Xp))::type>
+template <__cw_fixed_value _Xp, class = typename remove_cvref_t<decltype(_Xp)>::type>
 struct __constant_wrapper;
```
@davebayer is that correct?

miscco commented:
/ok to test [f3fe41c](https://github.com/NVIDIA/cccl/pull/5725/commits/f3fe41c804b8e507d84b0ea711c622974ca6eaf6)

alliepiper committed:
Merge remote-tracking branch 'origin/main' into windows_ci

# Conflicts:
#	ci/matrix.yaml

alliepiper committed:
Update override matrix.

alliepiper commented:
/ok to test f8d64c4

alliepiper committed:
Add placeholders for excluded c2h features.

suppress intentional "no tests ran" case

alliepiper committed:
Exclude cub NVRTC on MSVC

alliepiper committed:
Disable failing msvc test that's slated for refactor.

alliepiper commented:
/ok to test b7c331f

fbusato commented:
I'm not able to reproduce the libcu++ failure:
> libcu++ nvcc MSVC / EY / [CTK12.9 MSVC14.44 C++17] Test(amd64, RTX2080)
> complex_math_trig.pass.cpp

tested locally on CTK12.9 MSVC14.44 C++17, SM75 (RTX2080) and executed on SM89. All other configurations work in CI. 
I start thinking that can be an HW failure instead.

fbusato commented:
also, would make sense to restrict the target SM version for the compiler to `native` or the specific one to save CI resources and testing time

alliepiper committed:
[libcu++] Skip SIMD complex bfloat16 trig test on MSVC

alliepiper commented:
/ok to test 7ca584133a306b2c85190d27fe6be1c0bf651b73

alliepiper commented:
> also, would make sense to restrict the target SM version for the compiler to native or the specific one to save CI resources and testing time

The current implementation matches the existing build coverage conventions and just adds testing. Changing arch build coverage might make sense in some cases, but would be more appropriate for a follow-up.

alliepiper committed:
[CUB] Skip env-based deterministic reduce tests on MSVC + CTK 13.3

alliepiper committed:
Merge remote-tracking branch 'origin/main' into windows_ci

alliepiper commented:
/ok to test 6d1d8f3ba3dee798ad401066ae15dcbb31d1d376

alliepiper committed:
[CUB] Skip env-based reduce environment tests on MSVC + CTK 13.3

alliepiper commented:
/ok to test 6db84dde6a98b0dfa06e02a11e3284f0a21fa324

alliepiper committed:
[CI] Reset override matrix to empty

alliepiper commented:
Random external open-source contributor here, I think this is ready for another look 🙏

alliepiper commented:
Should pass once main is fixed.
